3.1 non root persistent lookback docs (#109)
logdna/logdna-agent-v2: docs/KUBERNETES.md

kubectl patch daemonset -n logdna-agent logdna-agent --type json -p '[{"op":"add","path":"/spec/template/spec/containers/0/securityContext/runAsUser","value":5000},{"op":"add","path":"/spec/template/spec/containers/0/securityContext/runAsGroup","value":5000}]'
```

### Enabling persistent agent state

> 

Reviewer: My only issue - will users get "enabling persistent agent state" = "not omitting/duplicating logs?"? Otherwise I'd think fine. Thought this was a issue whether running as root or non-root.

> To avoid skipping or duplicating logs when the agent pod restarts or is replaced the agent needs to store it's progress on the host node's filesystem. This is achieved using a hostPath volume. The host directory must be writable by the user or group specified in the securityContext.

Reviewer: again--"current state" ... sufficient?

> 

Reviewer: securityContext is just that...not an env variable. I'd say the yaml is a given.

To achieve this the host directory must either already exist with the correct
permissions or else Kubernetes will create the directory without write permissions for the agent user.

In this case the permissions must be set before the agent starts. When running as non-root the agent pod does not have permissions to do this, so an initcontainer may be used.

> 

Reviewer 2: Suggested rewrite of lines 194-197: "To achieve this the host directory must already exist with the correct permissions before the agent starts (otherwise Kubernetes will create the directory without the needed write permissions for the agent user). When running as non-root the agent pod does not have permissions to configure the host directory, so we use an *init container*, which is run before the app containers are started, to set the host directory permissions."

Reviewer: Not sure why "init container" would be italicized above. We'd need to do the same in other places in the doc. Also, I don't believe *we* use an init container - we're telling the user to use one

Reviewer 2: the use of italics is standard for the first time a new term is introduced. A way to call attention to it, make sure the reader sees it. So we would only use it once. BUt.. we can also drop it, or use `init container`.

Reviewer: Got that--but do we do the same for first use of other terms in the doc? That was my concern.

Below is an example manifest section which can be added to the pod specification alongside the containers array. It assumes the agent user/group are both 5000 and the volume is mounted at /var/lib/logdna

> 

Reviewer 2: can we say "...an example manifest section of the -file-name-.yaml which can be added..." just to again be uber-clear about **where** this section lives? If you think everyone knows, skip it, but... ;-)

Also can we add a few more words to explain what 5000 means.. is this a permission level? Like the top level or...? I tried googling it and found nothing.

Reviewer: I think the mention of "containers array"  is sufficient. 5000:5000 - just an example for setting ownership - user and group.

kubectl patch daemonset -n logdna-agent logdna-agent --type json -p '[{"op":"add","path":"/spec/template/spec/containers/0/securityContext/runAsUser","value":5000},{"op":"add","path":"/spec/template/spec/containers/0/securityContext/runAsGroup","value":5000}]'
```

### Enabling file offset tracking across restarts

> To avoid possible duplication or skipping of log messages during agent restart or upgrade, the agent stores its current file offsets on the host node's filesystem, using a hostPath volume.

Reviewer 3: Use host or node but not both.

```suggestion
To avoid possible duplication or skipping of log messages during agent restart or upgrade, the agent stores its current file offsets on the host's filesystem, using a `hostPath` volume.
```

> The host directory must be writable by the user or group specified in the securityContext.

Reviewer 3: ```suggestion
The host directory must be writable by the user or group specified in the `securityContext`.
```

To achieve this the host directory must either already exist with the correct
permissions or else Kubernetes will create the directory without write permissions for the agent user.

In this case the permissions must be set before the agent starts. When running as non-root the agent pod does not have permissions to do this, so an initcontainer may be used.

> Below is an example manifest section which can be added to the pod specification alongside the containers array. It assumes the agent user/group are both 5000 and the volume is mounted at /var/lib/logdna

Reviewer 3: ```suggestion
Below is an example manifest section which can be added to the pod specification alongside the containers array. It assumes the agent user/group are both `5000` and the volume is mounted at `/var/lib/logdna`:
```
